src/assets.zig

```diff
@@ -318,34 +417,78 @@ pub const Palette = struct { // MARK: Palette
 		}
 		return zon;
 	}
+
+	pub fn size(self: *Palette) usize {
+		return self.palette.items.len;
+	}
+
+	pub fn replaceEntry(self: *Palette, entryIndex: usize, newEntry: []const u8) void {
+		self.palette.allocator.free(self.palette.items[entryIndex]);
+		self.palette.items[entryIndex] = self.palette.allocator.dupe(u8, newEntry);
+	}
 };
 
 var loadedAssets: bool = false;
 
 pub fn loadWorldAssets(assetFolder: []const u8, blockPalette: *Palette, biomePalette: *Palette) !void { // MARK: loadWorldAssets()
 	if(loadedAssets) return; // The assets already got loaded by the server.
 	loadedAssets = true;
+
 	var blocks = commonBlocks.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
 	defer blocks.clearAndFree();
+
+	var blockMigrations = commonBlocksMigrations.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
+	defer blockMigrations.clearAndFree();
+
 	var items = commonItems.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
 	defer items.clearAndFree();
+
+	var itemsMigrations = commonItemsMigrations.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
+	defer itemsMigrations.clearAndFree();
+
 	var tools = commonTools.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
 	defer tools.clearAndFree();
+
 	var biomes = commonBiomes.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
 	defer biomes.clearAndFree();
+
+	var biomesMigrations = commonBiomesMigrations.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
+	defer biomesMigrations.clearAndFree();
+
 	var recipes = commonRecipes.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
 	defer recipes.clearAndFree();
+
 	var models = commonModels.cloneWithAllocator(main.stackAllocator.allocator) catch unreachable;
 	defer models.clearAndFree();
 
-	readAssets(arenaAllocator, assetFolder, &blocks, &items, &tools, &biomes, &recipes, &models);
+	readAssets(
+		arenaAllocator,
+		assetFolder,
+		&blocks,
+		&blockMigrations,
+		&items,
+		&itemsMigrations,
+		&tools,
+		&biomes,
+		&biomesMigrations,
+		&recipes,
+		&models,
+	);
 	errdefer unloadAssets();
 
-	var modelIterator = models.iterator();
-	while (modelIterator.next()) |entry| {
-		_ = main.models.registerModel(entry.key_ptr.*,  entry.value_ptr.*);
+	// Migrations have to be registered before the world is loaded so they can be
+	// applied to the world during load.
+	migrations_zig.registerBlockMigrations(&commonBlocksMigrations);
```

Comment:
I really like that part, the fact that this is just a simple function call, not a loop. I think this patter should be repeated for blocks, biomes etc, as asset loader is not really concerned with how those assets are structured. IMO each of the asset modules (what is the proper name?) like `blocks.zig` should have function like `registerBlocks` that would just take `*std.StringHashMap(ZonElement)` and do with it whatever is necessary. Right now we have mixed levels of abstraction - first we have high level function calls, but then we have lower level of abstraction where we manually iterate over data structure. I would be keen on draining `assets.zig` from everything that is not directly a asset loading implementation detail. Right now we have a lot of different things here.

Reply:
assets.zig is one of the first files I wrote, so yeah its definitely far from ideal.
I like the idea of one singular `registerBlocks` function, and it could also make the resetting code when leaving a world a easier.